runtime/near-vm-runner/Cargo.toml

```diff
@@ -46,6 +46,7 @@ near-crypto.workspace = true
 near-o11y = { workspace = true, optional = true }
 near-parameters.workspace = true
 near-primitives-core.workspace = true
+near-primitives.workspace = true
```

Comment:
`near-vm-runner` depending on `near-primitives` is a non-starter, unfortunately. This dependency edge precludes one of the larger projects we have (currently on the backburner) been working on: [limited replayability](https://github.com/near/nearcore/issues/8197).

Reply:
I do however agree that the dependency edge from `near-primitives` to `near-vm-runner` is also undesirable. I've been thinking that ideally anything relevant to near-vm-runner should only appear inside the crate that's integrating said code and that this exact code should be responsible for converting to & from the data structures (thus integrating the individual components.)

Reply:
> Well, currently, near-vm-runner breaks windows and don't allow to make a release :(

We've never really had windows on the list of platforms for which we support building. It wouldn't be a terrible idea to gain support for it, but so far it hasn't blocked releases.

I would love to gain more and wider platform support, but undoing work that's been made towards other projects is a good way to go about it.

> Can I move the errors to near-primitives-core, and compiled contract to near-primitives-core?

Part of the limited replayability effort is to not have anything from `near-*` dependencies appear in the public API surface of `near-vm-runner`. Moving the errors, or the `ContractRuntimeCache` to another `near-*` crate would be directly incompatible with that goal.

In my experience error-like thingies are actually pretty straightforward to deal with here. See e.g. `runtime/runtime/src/conversions.rs` for how I had approached it for some of the types. I basically defined copies of the types in the two locations that need to deal with those errors and added a function to convert between the two in some common ancestor crate that glues the two "components" together. Eventually these errors diverge: the versions of the type defined in the `near-vm-runner` are already protocol-version-agnostic, whereas the `near_primitives` versions are very sensitive to this due to being serialized to the chain.

I've fewer ideas on what to do about `ContractRuntimeCache`. I'm not sure why you're triggering this dependency edge in the first place, but at least `near_store::StoreContractRuntimeCache` could be removed entirely, especially if that's the only roadblock.

Reply:
Well, I believe the block happened at 0.22 with adding rustix:fs (that is only unix compatible). Also, near-vm-runner in 0.22 release has pulled wasmer-runtime-core-near dep, that is also not working on windows :(

near-cli-rs uses primitives and primitives pull near-vm-runner that pulls some crates that are incompatible with windows : (

I tried to remove near-primitives connection to near-vm-runner, but yeah, it uses the trait. :(

Reply:
I tried to follow your suggestion and here is my second potential implementation:
https://github.com/near/nearcore/compare/master...akorchyn:nearcore:second-try?expand=1

It's a bit scary cause it has two definitions of `ContractRuntimeCache` and it creates wrapper traits for converting trait A into trait B...

wdyt? I haven't finished and there is bunch of todos, but I would love to see your opinion if it can go :)

Reply:
Resolve as per discussion in zulip